apps/src/code-studio/initApp/project.js

```diff
@@ -960,6 +963,38 @@ var projects = (module.exports = {
       });
     });
   },
+
+  /**
+   * Tests whether provided sample code is different from the current project code.
+   * This also normalizes the code so incidental differences in line endings or
+   * empty xml tags are not recognized as differences.
+   * @param {string} sampleCode the code to diff against the current project code
+   */
+  isCodeDifferent(sampleCode) {
+    let normalizedSample, normalizedCurrent;
+    const parser = new DOMParser();
+    const parsedCurrent = parser.parseFromString(
+      currentSources.source,
+      'text/xml'
+    );
+    const parsedSample = parser.parseFromString(sampleCode, 'text/xml');
+    if (
+      parsedCurrent.getElementsByTagName('parsererror').length > 0 ||
+      parsedSample.getElementsByTagName('parsererror').length > 0
+    ) {
+      // This is droplet. Normalize line endings.
```

Comment:
For my education: Why do we need to normalize line endings when it's droplet?

Reply:
Droplet is just a fancy rendering of javascript. The text box that displays the code seems to be adjusting the line endings based on the user's operating system. I think my comment may be confusing things. Rather than "This is droplet, therefore we should normalize line endings" it's supposed to be "FYI: droplet code will go through this branch of the condition" and "Next step: normalize line endings"